src/codegen/transformations.lisp

```diff
@@ -123,3 +124,37 @@ both CL namespaces appearing in `node`"
            (type parser:identifier-list bound-variables)
            (values boolean &optional))
   (null (intersection (node-variables node) bound-variables)))
+
+(defun node-returns-p (node)
+  "Does `node` contain a `node-return` subnode?"
+  (declare (type node node)
+           (values boolean &optional))
+  (traverse
+   node
+   (list
+    (action (:before node-return _node)
+      (declare (ignore _node))
+      (return-from node-returns-p t))))
+  nil)
+
+(defun localize-returns (node)
+  "If `node` contains any `node-return` subnodes, wrap `node` in a new
+`node-block` and rewrite all `node-return` subnodes to return from
+this block."
+  (declare (type node node)
+           (values node &optional))
+  (unless (node-returns-p node)
+    (return-from localize-returns node))
+
+  (let ((block-name (gentemp)))
```

Comment:
Should I make the inliner pass a function name so the block name can be something more informative like `INLINED-FUNCTIONNAME` or `INLINED-LAMBDA`?

Reply:
we might do an earlier pass altogether that transforms ALL functions to always have a block around the contents. thoughts?

Reply:
I think this is an interesting possibility, and I tried to implement this in the tc-node -> ast-node translation. The sad thing is that the output of the codegen will then look messier with these extra blocks instead, of return-from function-name, unless there is extra logic added to codegen-expression